packages/flutter/lib/src/material/menu_bar.dart

```diff
@@ -0,0 +1,3256 @@
+    required List<MenuItem> topLevel,
+    required bool enabled,
+  }) {
+    menuBarContext = context;
+    _createMenuTree(topLevel);
+    if (!_listeningToPointerEvents) {
+      GestureBinding.instance.pointerRouter.addGlobalRoute(_handlePointerEvent);
+      _listeningToPointerEvents = true;
+    }
+
+    this.enabled = enabled;
+
+    // Need to mark dirty in a post frame callback, since this update might
+    // updated part of the tree that isn't in the overlay, but calling this
+    // would request that the overlay be rebuilt.
+    SchedulerBinding.instance.addPostFrameCallback((Duration _) {
+      _markMenuDirty();
+    });
+  }
+
+  // Disconnects this controller from the menu bar, in preparation for disposal,
+  // or when the controller is being swapped out for another one.
+  void disconnect() {
+    _menuBarContext = null;
+    root.children.clear();
+    _focusNodes.clear();
+    _previousFocus = null;
+    _focusBeforeClick = null;
+    _pendingFocusedMenu = null;
+    if (_listeningToPointerEvents) {
+      GestureBinding.instance.pointerRouter.removeGlobalRoute(_handlePointerEvent);
+      _listeningToPointerEvents = false;
+    }
+  }
+
+  @override
+  void dispose() {
+    disconnect();
+    menuBarScope.dispose();
+    overlayScope.dispose();
+    super.dispose();
+    _disposed = true;
+  }
+
+  /// Closes the given menu, and any open descendant menus.
+  ///
+  /// Leaves ancestor menus open, if any.
+  ///
+  /// Notifies listeners if the menu changed.
+  void close(_MenuNode node) {
+    if (openMenu == null) {
+      // Everything is already closed.
+      return;
+    }
+    if (isAnOpenMenu(node)) {
+      // Don't call onClose, notifyListeners, etc, here, because setting
+      // openMenu will call them if needed.
+      if (node.parent == root) {
+        openMenu = null;
+      } else {
+        openMenu = node.parent;
+      }
+    }
+  }
+
+  @override
+  void closeAll() {
+    openMenu = null;
+  }
+
+  /// Returns true if the given menu or one of its ancestors is open.
+  bool isAnOpenMenu(_MenuNode menu) {
+    if (_openMenu == null) {
+      return false;
+    }
+    return _openMenu == menu || (_openMenu?.ancestors.contains(menu) ?? false);
+  }
+
+  // Handles any pointer events that occur in the app, checking them against
+  // open menus to see if the menus should be closed or not.
+  // This isn't called if no menus are open.
+  void _handlePointerEvent(PointerEvent event) {
+    if (event is! PointerDownEvent) {
+      return;
+    }
+    bool isInsideMenu = false;
+    final List<RenderBox> renderBoxes = <RenderBox?>[
+      menuBarContext.findRenderObject() as RenderBox?,
+      ..._menuRenderBoxes,
+    ].where((RenderBox? box) => box != null).cast<RenderBox>().toList();
+    for (final RenderBox renderBox in renderBoxes) {
+      assert(renderBox.attached);
+      isInsideMenu =
+          renderBox.hitTest(BoxHitTestResult(), position: renderBox.globalToLocal(event.position)) || isInsideMenu;
```

Comment:
PointerEvent has a localPosition getter. Does that work?

Reply:
Oh, it probably doesn't because this is a global route and therefore the position wouldn't have been transformed correctly to the renderBox. Nevermind.